### PR TITLE
[jit_ext] Inserts unpack bsyms to traces generated inside `torch.autograd.Function`

### LINKED ISSUE #1413: `dce` inside `torch.autograd.Function` could fail as its traces miss unpacking bsyms
This is the line in question: https://github.com/Lightning-AI/lightning-thunder/blob/a0a61518a70d80ab5ce4d1b6e44bcae159d81bb0/thunder/core/transform_common.py#L140

When I tried `dce`ing the following trace, the keys of the `producer_map` inside `dce` are `tuple(producer_map._dict.keys()) = ('t15', 't23', 't24', 't25', 't26')`, leading to 
```
Encountered exception KeyError: 't_0'
```

```python
def _ToFloat8ConstrFunc(t_0, t12, p27, tup28, INPUT, p29):
  # /home/mkozuki/ghq/github.com/crcrpar/ao/torchao/float8/float8_tensor.py:170:                tensor_scaled = tensor.to(torch.float32) * scale
  t_0 = ltorch.to(t_0, torch.float32, None, device=None, dtype=None, copy=False, memory_format=None)  # t_0: "cuda:0 f32[16, 32]"
  t15 = ltorch.mul(t_0, t12)  # t15: "cuda:0 f32[16, 32]"
    # t14 = prims.broadcast_in_dim(t12, (16, 32), ())  # t14: "cuda:0 f32[16, 32]"
    # t15 = prims.mul(t_0, t14)  # t15: "cuda:0 f32[16, 32]"

  # /home/mkozuki/ghq/github.com/crcrpar/ao/torchao/float8/float8_utils.py:161:                 x = x.clamp(min=-max_value, max=max_value)
  t23 = ltorch.clamp(t15, -448.0, 448.0)  # t23: "cuda:0 f32[16, 32]"
    # t16 = ltorch.ne(t15, t15)  # t16: "cuda:0 b8[16, 32]"
      # t16 = prims.ne(t15, t15)  # t16: "cuda:0 b8[16, 32]"
    # t17 = ltorch.gt(t15, -448.0)  # t17: "cuda:0 b8[16, 32]"
      # t17 = prims.gt(t15, -448.0)  # t17: "cuda:0 b8[16, 32]"
    # t18 = ltorch.where(t17, t15, -448.0)  # t18: "cuda:0 f32[16, 32]"
      # t18 = prims.where(t17, t15, -448.0)  # t18: "cuda:0 f32[16, 32]"
    # t19 = ltorch.where(t16, t15, t18)  # t19: "cuda:0 f32[16, 32]"
      # t19 = prims.where(t16, t15, t18)  # t19: "cuda:0 f32[16, 32]"
    # t20 = ltorch.ne(t19, t19)  # t20: "cuda:0 b8[16, 32]"
      # t20 = prims.ne(t19, t19)  # t20: "cuda:0 b8[16, 32]"
    # t21 = ltorch.lt(t19, 448.0)  # t21: "cuda:0 b8[16, 32]"
      # t21 = prims.lt(t19, 448.0)  # t21: "cuda:0 b8[16, 32]"
    # t22 = ltorch.where(t21, t19, 448.0)  # t22: "cuda:0 f32[16, 32]"
      # t22 = prims.where(t21, t19, 448.0)  # t22: "cuda:0 f32[16, 32]"
    # t23 = ltorch.where(t20, t19, t22)  # t23: "cuda:0 f32[16, 32]"
      # t23 = prims.where(t20, t19, t22)  # t23: "cuda:0 f32[16, 32]"

  # /home/mkozuki/ghq/github.com/crcrpar/ao/torchao/float8/float8_utils.py:162:                 return x.to(float8_dtype)
  t24 = ltorch.to(t23, torch.float8_e4m3fn, None, device=None, dtype=None, copy=False, memory_format=None)  # t24: "cuda:0 f8_e4m3fn[16, 32]"
    # t24 = prims.convert_element_type(t23, dtypes.float8_e4m3fn)  # t24: "cuda:0 f8_e4m3fn[16, 32]"

  # /home/mkozuki/ghq/github.com/crcrpar/ao/torchao/float8/float8_tensor.py:318:                    data.size(),
  (_, _) = prims.shape(t24)

  # /home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/thunder/core/proxies.py:1957:                   self.requires_grad,
  t25 = prims.tensor_subclass_ctor(_torch__C__TensorMeta_0, 't25', (16, 32), devices.Device("cuda:0"), dtypes.float32, False, [t24, t12], [dtypes.float32, None, _GemmInputRole_1, False, True, False, False, False, False, False, False, False, False, False, False])  # t25: "cuda:0 f32[16, 32]"
  t26 = prims.shallow_copy(t25)  # t26: "cuda:0 f32[16, 32]"
  return t26
```

See https://gist.github.com/crcrpar/787f7241744be313a8d24f6efe0a8826 for more details

### PR DESCRIPTION
**Pull request by crcrpar**
## What does this PR do?

Fixes #1413.

### CODE REVIEW

File: thunder/core/jit_ext.py [hunk 1]
```diff
@@ -680,6 +681,12 @@ def _general_jit_torch_autograd_function_apply_lookaside(obj: Any, *args, **kwar
         unwrapped_custom_forward_args,
     )
     trace_of_fwd.args = unwrapped_custom_forward_args
+    unpack_bsyms = [
+        prims.unpack_trivial.bind(a, name=a.name, output=a)
+        for a in filter(lambda a: isinstance(a, Proxy), trace_of_fwd.args)
+    ]
+    trace_of_fwd.bound_symbols = unpack_bsyms + trace_of_fwd.bound_symbols
+    trace_of_fwd = dce(trace_of_fwd)
```
Comment (t-vi) [inline thunder/core/jit_ext.py:689]: Do you hit errors without the DCE? It'd be cool to keep the DCE until after the initial trace.

Comment (crcrpar) [reply thunder/core/jit_ext.py:689]: let me narrow down the scope of this pr to just correctly propagate exceptions

**Review by t-vi — APPROVED** [review]
Thank you @crcrpar

File: thunder/core/jit_ext.py (hunk 1 above)
Comment (t-vi) [reply thunder/core/jit_ext.py]: Thank you.

**Pull request merged**